File: Leetcode/PY/Leetcode53-最大子序和.py
from typing import List
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        ans,n = nums[0],len(nums)
        dp = [0] * 2
        dp[0] = nums[0]
        old,now = 0,0
        for i in range(1,n):
            old = now
            now = 1- now
            dp[now] = max(nums[i],dp[old] + nums[i])
            ans = max(ans,dp[now])
        return ans

File: Leetcode/Rust/Leetcode53-最大子序和.rs
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let (mut n,mut ans) = (nums.len(),nums[0]);
        let mut dp = vec![0;2];
        dp[0] = nums[0];
        let (mut old,mut now) = (0,0);
        for i in 1..n{
            old = now;
            now = 1 - now;
            dp[now] = nums[i].max(dp[old] + nums[i]);
            ans = ans.max(dp[now])
        } 
        ans
    }
}